test/test_kernel_launch.jl

```diff
@@ -0,0 +1,138 @@
+include("common.jl")
+
+import Chmy.Architectures: deepmap!, disable_task_sync!, enable_task_sync!
+
+struct CuArray end # in reality this is replaced by "using CUDA"
+
+myfun!(::Any) = nothing
+
+for backend in TEST_BACKENDS, T in TEST_TYPES
+    if !compatible(backend, T)
+        continue
+    end
+
+    @testset "$(basename(@__FILE__)) (backend: $backend, type: $T)" begin
+        if (backend isa CPU)
+            @testset "deepmap! on CPU backend" begin
+                # struct CuArray end # in reality this is replaced by "using CUDA"
+
+                calls = Symbol[]
+
+                disable_task_sync!(::Any) = nothing
+                enable_task_sync!(::Any) = nothing
+                disable_task_sync!(::CuArray) = push!(calls, :disable) # actual unsafe_disable_task_sync! call
+                enable_task_sync!(::CuArray) = push!(calls, :enable) # actual unsafe_enable_task_sync! call
+
+                empty!(calls)
+                foo1 = (1, 2, 3)
+                deepmap!(disable_task_sync!, foo1) # prints nothing
+                deepmap!(enable_task_sync!, foo1)  # prints nothing
+                @test isempty(calls)
+
+                empty!(calls)
+                foo2 = ()
+                deepmap!(disable_task_sync!, foo2) # prints nothing
+                deepmap!(enable_task_sync!, foo2)  # prints nothing
+                @test isempty(calls)
+
+                empty!(calls)
+                foo3 = (CuArray(), CuArray(), (1, ()))
+                deepmap!(disable_task_sync!, foo3) # prints "disable" twice
+                deepmap!(enable_task_sync!, foo3) # prints "enable" twice
+                @test calls == [:disable, :disable, :enable, :enable]
+
+                empty!(calls)
+                foo4 = (a=1, b=CuArray(), c=((), 1))
+                deepmap!(disable_task_sync!, foo4) # prints "disable" once
+                deepmap!(enable_task_sync!, foo4) # prints "enable" once
+                @test calls == [:disable, :enable]
+
+                empty!(calls)
+                struct Foo{T}
+                    a::Int
+                    b::CuArray
+                    c::T
+                end
+                foo5 = Foo(1, CuArray(), foo4)
+
+                with_no_task_sync!(foo5) do
+                    push!(calls, :in_out) # kernel execution for inner and outer regions goes here
+                end # will print "disable" twice, then "in_out" and "enable" twice
+                @show calls
```

Comment:
It fails here as it looks like `with_no_task_sync!` seems not to call the `deepmap` which would dispatch on `CuArray` but on `Any` which returns `nothing` instead of pushing the symbol.